src/app/AttributeAccessInterface.h

```diff
@@ -51,7 +56,31 @@ class AttributeValueEncoder
         {
             return CHIP_NO_ERROR;
         }
-        return DataModel::Encode(*mWriter, TLV::ContextTag(AttributeDataElement::kCsTag_Data), std::forward<Ts>(aArgs)...);
+        return TagBoundTLVEncoder::Encode(std::forward<Ts>(aArgs)...);
+    }
+
+    /**
+     * aCallback is expected to take a const TagBoundTLVEncoder& argument and
+     * Encode() on it as many times as needed to encode all the list elements
+     * one by one.  If any of those Encode() calls returns failure, aCallback
+     * must stop encoding and return failure.  When all items are encoded
+     * aCallback is expected to return success.
+     *
+     * aCallback may not be called.  Consumers must not assume it will be
+     * called.
+     */
+    template <typename ListGenerator>
+    CHIP_ERROR EncodeList(ListGenerator aCallback)
```

Comment:
While you're at it, do you think it's useful to add an `EncodeStruct` too as well?

Reply:
That would require the caller to do something about tags; we can't really infer the tags for the struct items the way we know everything in a list is annonymous-tagged....

Reply:
Sure - but it avoids them having to store the type of the container on entry/exit. That would be the only win.

Reply:
Could go either one FYI.